### PR TITLE
Fix ORC schema visitors to support reading ORC files with deeply nest…

### PR DESCRIPTION
**Pull request by rzhang10**
…ed union type schema

### CODE REVIEW

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java [hunk 1]
```diff
@@ -123,8 +122,8 @@
     return visitor.record(record, names, visitFields(fields, names, visitor));
   }
 
-  public String optionName() {
-    return "_option";
+  public String optionName(int ordinal) {
+    return "tag_" + ordinal;
```
Comment (funcheetah) [inline orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java:126]: Could you give an example how this is used?

**Comment by funcheetah** [comment]
Thanks @rzhang10 for the quick fix.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java [hunk 2]
```diff
@@ -215,4 +216,83 @@ public void testSingleComponentUnion() throws IOException {
...
+        innerInt1.fillWithNulls();
+        innerString3.fillWithNulls();
+      }
+    }
+    if (batch.size != 0) {
+      writer.addRowBatch(batch);
+      batch.reset();
+    }
+    writer.close();
+
```
Comment (funcheetah) [inline spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:286]: Could you also add a test for `VectorizedSparkOrcReaders`?

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java (hunk 1 above)
Comment (rzhang10) [reply orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java:126]: It's used by the `ApplyNameMapping` visitor to correctly recognize the union converted struct field names, for example sth like `c1.tag_1.c2` and assign the correct ID to such field.

Comment (shenodaguirguis) [reply orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java:126]: I didn't check the ApplyNameMapping visitor, but can you verify that using same ordinal across unions - i.e., (1, "tag_0") and again (5, "tag_0") as in your test case -  could lead to any issues?

**Review by shenodaguirguis — APPROVED** [review]
LGTM, left one question to verify

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java (hunk 2 above)
Comment (rzhang10) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:286]: Added

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java (hunk 1 above)
Comment (rzhang10) [reply orc/src/main/java/org/apache/iceberg/orc/OrcSchemaVisitor.java:126]: It won't, the (1, tag_0) and (5, tag_0) are two fields at different nested levels inside the schema so there's no ambiguity.

**Review by funcheetah — APPROVED** [review]

**Pull request merged**